tools/xkbcli.1
.Dd June 4, 2024
.Dt XKBCLI 1
.Os
.
.Sh NAME
.Nm xkbcli
.Nd tool to interact with X Keyboard keymaps
.
.Sh SYNOPSIS
.Nm
.Ar command Bo arguments Bc
.
.Nm
.Op Fl \-help | Fl \-version
.
.Sh DESCRIPTION
.Nm
is a commandline tool to query, compile and test XKB keymaps, layouts and other elements.
.
.Bl -tag -width Ds
.It Fl h, \-help
Print help and exit
.
.It Fl V, \-version
Print the version and exit
.El
.
.Ss COMMANDS
.Bl -tag -width Ds
.It Ic info
Print information about xkbcommon configuration
.It Ic compile\-keymap
Compile an XKB keymap, see
.Xr xkbcli\-compile\-keymap 1
.
.It Ic compile\-compose
Compile a compose file, see
.Xr xkbcli\-compile\-compose 1
.
.It Ic how\-to\-type
Show how to type a given Unicode codepoint, see
.Xr xkbcli\-how\-to\-type 1
.
.It Ic interactive
Interactive debugger for XKB keymaps that automatically select from the
following backends, if available: Wayland, X11 and evdev.
See the corresponding entries hereinafter.
.
.It Ic interactive\-evdev
Interactive debugger for XKB keymaps for evdev, see
.Xr xkbcli\-interactive\-evdev 1
.
.It Ic interactive\-wayland
Interactive debugger for XKB keymaps for Wayland, see
.Xr xkbcli\-interactive\-wayland 1
.
.It Ic interactive\-x11
Interactive debugger for XKB keymaps for X11, see
.Xr xkbcli\-interactive\-x11 1
.
.It Ic dump\-keymap
Dump an XKB keymap from a Wayland or X11 compositor, see the corresponding
entries hereinafter.
.
.It Ic dump\-keymap\-wayland
Dump an XKB keymap from a Wayland compositor, see
.Xr xkbcli\-dump\-keymap\-wayland 1
.
.It Ic dump\-keymap\-x11
Dump an XKB keymap from an X server, see
.Xr xkbcli\-dump\-keymap\-x11 1
.
.It Ic list
List available layouts and more, see
.Xr xkbcli\-list 1
.El
.
.Pp
Note that not all tools may be available on your system.
.
.Sh EXIT STATUS
.Bl -tag -compact -width Ds
.It 0
exited successfully
.It 1
an error occured
.It 2
program was called with invalid arguments
.El
.
.Sh SEE ALSO
.Lk https://xkbcommon.org "The xkbcommon online documentation"


NEWS
libxkbcommon 1.6.0 - 2023-10-08
==================

API
---

### Breaking changes

- *Remove* keysyms that were intended for German T3 layout but are unused:

  - `XKB_KEY_dead_lowline`
  - `XKB_KEY_dead_aboveverticalline`
  - `XKB_KEY_dead_belowverticalline`
  - `XKB_KEY_dead_longsolidusoverlay`

  See the upstream [`xorgproto` MR](https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/70). See hereinafter for further changes related to keysyms.

### New

- Add Compose iterator API to iterate the entries in a compose table:

  - `xkb_compose_table_entry_sequence`
  - `xkb_compose_table_entry_keysym`
  - `xkb_compose_table_entry_utf8`
  - `xkb_compose_table_iterator_new`
  - `xkb_compose_table_iterator_free`
  - `xkb_compose_table_iterator_next`

- *Structured log messages* with a message registry. There is an *ongoing* work
  to assign unique identifiers to log messages and add a corresponding error
  index documentation page:

  - The log entries are preceded with an identifier in the form `XKB-NNN`, where
    `NNN` is a decimal number.

  - The log entries can then be parsed with third-party tools, to check for
    specific identifiers.

  - The new documentation page “**Error index**” lists all the kind of error messages
    with their identifiers. The aim is that each entry could present detailed
    information on the error and how to fix it.

- Add a new warning for numeric keysyms references in XKB files: the preferred
  keysym reference form is its name or its Unicode value, if relevant.

- Add the upper bound `XKB_KEYSYM_MAX` to check valid keysyms.

- Add a warning when loading a keymap using RMLVO with no layout but with the
  variant set. The variant is actually discarded and both layout and variant are
  set to default values, but this was done previously with no warning, confusing
  end users.

- Add support for `modifier_map None { … }`. This feature is missing compared to
  the X11 implementation. It allows to reset the modifier map of a key.

- Update keysyms using latest [xorgproto] (commit: `1c8128d72df22843a2022576850bc5ab5e3a46ea`):

  - Additions:

    - `XKB_KEY_ordmasculine` ([xorgproto-68])
    - `XKB_KEY_guillemetleft` ([xorgproto-68])
    - `XKB_KEY_guillemetright` ([xorgproto-68])
    - `XKB_KEY_dead_schwa` ([xorgproto-78])
    - `XKB_KEY_dead_SCHWA` ([xorgproto-78])
    - `XKB_KEY_dead_hamza` ([xorgproto-71])
    - `XKB_KEY_XF86EmojiPicker` ([xorgproto-44])
    - `XKB_KEY_XF86Dictate` ([xorgproto-49])
    - `XKB_KEY_XF86CameraAccessEnable` ([xorgproto-82])
    - `XKB_KEY_XF86CameraAccessDisable` ([xorgproto-82])
    - `XKB_KEY_XF86CameraAccessToggle` ([xorgproto-82])
    - `XKB_KEY_XF86NextElement` ([xorgproto-82])
    - `XKB_KEY_XF86PreviousElement` ([xorgproto-82])
    - `XKB_KEY_XF86AutopilotEngageToggle` ([xorgproto-82])
    - `XKB_KEY_XF86MarkWaypoint` ([xorgproto-82])
    - `XKB_KEY_XF86Sos` ([xorgproto-82])
    - `XKB_KEY_XF86NavChart` ([xorgproto-82])
    - `XKB_KEY_XF86FishingChart` ([xorgproto-82])
    - `XKB_KEY_XF86SingleRangeRadar` ([xorgproto-82])
    - `XKB_KEY_XF86DualRangeRadar` ([xorgproto-82])
    - `XKB_KEY_XF86RadarOverlay` ([xorgproto-82])
    - `XKB_KEY_XF86TraditionalSonar` ([xorgproto-82])
    - `XKB_KEY_XF86ClearvuSonar` ([xorgproto-82])
    - `XKB_KEY_XF86SidevuSonar` ([xorgproto-82])
    - `XKB_KEY_XF86NavInfo` ([xorgproto-82])

  - Deprecations:

    - `XKB_KEY_masculine`: use `XKB_KEY_ordmasculine` instead ([xorgproto-68])
    - `XKB_KEY_guillemotleft`: use `XKB_KEY_guillemetleft` instead ([xorgproto-68])
    - `XKB_KEY_guillemotright`: use `XKB_KEY_guillemetright` instead ([xorgproto-68])
    - `XKB_KEY_dead_small_schwa`: use `XKB_KEY_dead_schwa` instead ([xorgproto-78])
    - `XKB_KEY_dead_capital_schwa`: use `XKB_KEY_dead_SCHWA` instead ([xorgproto-78])

  [xorgproto]: https://gitlab.freedesktop.org/xorg/proto/xorgproto
  [xorgproto-44]: https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/44
  [xorgproto-49]: https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/49
  [xorgproto-68]: https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/68
  [xorgproto-71]: https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/71
  [xorgproto-78]: https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/78
  [xorgproto-82]: https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/82

- Ongoing work to improve the documentation about XKB and its V1 format.

### Fixes

- Prevent `xkb_keysym_from_name` to parse out-of-range hexadecimal keysyms.

- Disallow producing NULL character with escape sequences `\0` and `\x0`.

- Prevent overflow of octal escape sequences by making `\400..\777` invalid.

- Prevent interpreting and emitting invalid Unicode encoding forms: surrogates
  are invalid in both UTF-32 and UTF-8.

- `xkb_keymap_new_from_buffer`: Allow for a NULL-terminated keymap string.

- Compose: Increase the limit of possible entries to handle huge Compose files.

  Contributed by alois31.

Tools
-----

### New

- Add bash completions for `xkbcli` and its subcommands.

- `xkbcli interactive-*`: Add options `--short` to hide some fields.

- `xkbcli interactive-evdev`: Add `--includes` and `--include-defaults` options.

- Add `xkb-check-messages` *experimental* tool (not installed).
  It checks whether given log messages identifiers are supported.

### Fixes

- `xkbcli compile-keymap`: Allow to use without arguments.

- `xkbcli interactive-*`: Always print keycode.

- `xkbcli interactive-*`: Escape control characters for Unicode output, instead of
  printing them as-is, messing the output.

Build system
------------

- Bump required meson to 0.52.0.

- Allow `xkbcommon` to be used as a subproject.

  Contributed by Simon Ser.

- Improve Windows compilation.

libxkbcommon 1.5.0 - 2023-01-02
==================

- Add `xkb_context` flag `XKB_CONTEXT_NO_SECURE_GETENV` and `rxkb_context` flag
  `RXKB_CONTEXT_NO_SECURE_GETENV`.


tools/xkbcli.c
/*
 * Copyright © 2020 Red Hat, Inc.
 * SPDX-License-Identifier: MIT
 */

#include "config.h"

#include <getopt.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>

#include "tools-common.h"

static void
usage(void)
{
    printf("Usage: xkbcli [--help|-h] [--version|-V] <command> [<args>]\n"
           "\n"
           "Global options:\n"
           "  -h, --help ...... show this help and exit\n"
           "  -V, --version ... show version information and exit\n"
           /* WARNING: The following is parsed by the bash completion script.
            *          Any change to the format (in particular to the indentation)
            *          should kept in the script in sync. */
           "Commands:\n"
#if HAVE_XKBCLI_INFO
           "  info\n"
           "    Print information about xkbcommon configuration\n"
           "\n"
#endif
#if HAVE_XKBCLI_LIST
           "  list\n"
           "    List available rules, models, layouts, variants and options\n"
           "\n"
#endif
#if HAVE_XKBCLI_INTERACTIVE_WAYLAND || HAVE_XKBCLI_INTERACTIVE_X11
           "  interactive\n"
           "    Interactive debugger for XKB keymaps; automatically select from"
           "    the following backends, if available: Wayland, X11 and evdev.\n"
           "\n"
#endif
#if HAVE_XKBCLI_INTERACTIVE_WAYLAND
           "  interactive-wayland\n"
           "    Interactive debugger for XKB keymaps for Wayland\n"
           "\n"
#endif
#if HAVE_XKBCLI_INTERACTIVE_X11
           "  interactive-x11\n"
           "    Interactive debugger for XKB keymaps for X11\n"
           "\n"
#endif
#if HAVE_XKBCLI_INTERACTIVE_EVDEV
           "  interactive-evdev\n"
           "    Interactive debugger for XKB keymaps for evdev\n"
           "\n"
#endif
#if HAVE_XKBCLI_DUMP_KEYMAP_WAYLAND || HAVE_XKBCLI_DUMP_KEYMAP_X11
           "  dump-keymap\n"
           "    Dump an XKB keymap from a Wayland or X11 compositor\n"
           "\n"
#endif
#if HAVE_XKBCLI_DUMP_KEYMAP_WAYLAND
           "  dump-keymap-wayland\n"
           "    Dump an XKB keymap from a Wayland compositor\n"
           "\n"
#endif
#if HAVE_XKBCLI_DUMP_KEYMAP_X11
           "  dump-keymap-x11\n"
           "    Dump an XKB keymap from an X server\n"
           "\n"
#endif
#if HAVE_XKBCLI_COMPILE_KEYMAP
           "  compile-keymap\n"
           "    Compile an XKB keymap\n"
           "\n"
#endif
#if HAVE_XKBCLI_COMPILE_COMPOSE
           "  compile-compose\n"
           "    Compile a Compose file\n"
           "\n"
#endif
#if HAVE_XKBCLI_HOW_TO_TYPE
           "  how-to-type\n"
           "    Print key sequences to type a Unicode codepoint\n"
           "\n"
#endif
           );
}

int
main(int argc, char **argv)
{
    enum options {
        OPT_HELP = 1,
        OPT_VERSION,
    };
    int option_index = 0;

    setlocale(LC_ALL, "");

    while (1) {
        int c;
        static struct option opts[] = {
            { "help",    no_argument, 0, OPT_HELP },
            { "version", no_argument, 0, OPT_VERSION },
            { 0, 0, 0, 0}
        };

        c = getopt_long(argc, argv, "+hV", opts, &option_index);
        if (c == -1)
            break;

        switch(c) {
            case 'h':
            case OPT_HELP:
                usage();
                return EXIT_SUCCESS;
            case 'V':
            case OPT_VERSION:
                printf("%s\n", LIBXKBCOMMON_VERSION);
                return EXIT_SUCCESS;
            default:
                usage();
                return EXIT_INVALID_USAGE;
        }
    }

    if (optind >= argc) {
        usage();
        return EXIT_INVALID_USAGE;
    }

    argv += optind;
    argc -= optind;

    return tools_exec_command("xkbcli", argc, (const char **) argv);
}
